test/cpp/util/channelz_sampler_test.cc

```diff
@@ -0,0 +1,172 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/grpc.h>
+#include <grpc/support/port_platform.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/ext/channelz_service_plugin.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/security/credentials.h>
+#include <grpcpp/security/server_credentials.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+#include <grpcpp/server_context.h>
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include "include/grpcpp/ext/proto_server_reflection_plugin_impl.h"
+#include "src/cpp/server/channelz/channelz_service.h"
+#include "src/proto/grpc/testing/test.grpc.pb.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/util/subprocess.h"
+#include "test/cpp/util/test_credentials_provider.h"
+
+using grpc::Channel;
+using grpc::ClientContext;
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::Status;
+
+#include <gtest/gtest.h>
+
+// Test variables
+std::string server_address("0.0.0.0:10000");
+std::string custom_credentials_type("INSECURE_CREDENTIALS");
+std::string sampling_times = "2";
+std::string sampling_interval_seconds = "5";
+std::string output_json("./output.json");
+
+// Creata an echo server - randomly delay 0.1 to 0.2 s
+class EchoServerImpl final : public grpc::testing::TestService::Service {
+  Status EmptyCall(::grpc::ServerContext* context,
+                   const grpc::testing::Empty* request,
+                   grpc::testing::Empty* response) {
+    srand(unsigned(time(0)));
+    unsigned int server_delay_microseconds = 100000;
+    server_delay_microseconds += rand() % server_delay_microseconds;
+    usleep(server_delay_microseconds);
+    return Status::OK;
+  }
+};
+
+// Run server in a thread
+void RunServer() {
+  // register channelz service
+  ::grpc::channelz::experimental::InitChannelzService();
+
+  EchoServerImpl service;
+  grpc::EnableDefaultHealthCheckService(true);
+  grpc_impl::reflection::InitProtoReflectionServerBuilderPlugin();
+  grpc::ServerBuilder builder;
+  auto server_creds =
+      grpc::testing::GetCredentialsProvider()->GetServerCredentials(
+          custom_credentials_type);
+  builder.AddListeningPort(server_address, server_creds);
+
+  // forces channelz and channel tracing to be enabled.
+  builder.AddChannelArgument(GRPC_ARG_ENABLE_CHANNELZ, 1);
+  builder.AddChannelArgument(GRPC_ARG_MAX_CHANNEL_TRACE_EVENT_MEMORY_PER_NODE,
+                             1024);
+  builder.RegisterService(&service);
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  gpr_log(GPR_INFO, "Server listening on %s", server_address.c_str());
+  server->Wait();
+}
+
+// Creata an echo client - set timeout as 0.15s
+class EchoClientImpl {
```

Comment:
nit: I don't see a lot of advantage to this being a class, so I suggest removing this, and inlining the fields and logic of this class into the `RunClient` loop.

Reply:
Removed